src/Annotation/AbstractBuilder.php

```diff
@@ -0,0 +1,353 @@
+<?php
+
+namespace Laminas\Form\Annotation;
+
+use ArrayObject;
+use Laminas\EventManager\Event;
+use Laminas\EventManager\EventManager;
+use Laminas\EventManager\EventManagerAwareInterface;
+use Laminas\EventManager\EventManagerInterface;
+use Laminas\Form\Element;
+use Laminas\Form\Exception;
+use Laminas\Form\Factory;
+use Laminas\Form\FieldsetInterface;
+use Laminas\Form\FormFactoryAwareInterface;
+use Laminas\Form\FormInterface;
+use Laminas\Stdlib\ArrayUtils;
+use ReflectionClass;
+use ReflectionProperty;
+use Reflector;
+
+use function get_class;
+use function is_string;
+use function is_subclass_of;
+
+/**
+ * Creates a form and input-filters from an array-based form specification
+ */
+abstract class AbstractBuilder implements EventManagerAwareInterface, FormFactoryAwareInterface
+{
+    /**
+     * @var EventManagerInterface
+     */
+    protected $events;
+
+    /**
+     * @var Factory
+     */
+    protected $formFactory;
+
+    /**
+     * @var object
+     */
+    protected $entity;
+
+    /**
+     * @var bool
+     */
+    protected $preserveDefinedOrder = false;
+
+    /**
+     * Set form factory to use when building form from annotations
+     *
+     * @param  Factory $formFactory
+     * @return $this
+     */
+    public function setFormFactory(Factory $formFactory)
```

Comment:
@Slamdunk 
What is your current plan? Do you want to release another major version to add the return types or is this also a topic for version 3?

Reply:
I'm sorry I haven't documented it because I didn't expect @driehle doing such great PRs and postponing my job about types ^^
It is my intention for `v3` to have `laminas-codingstandard:v2` and reboot PSalm to have empty baseline and incrementally raise its level from 8 to whatever I can reach, and in the path I'll add types (or move them from DocBlock to code) everywhere.

Reply:
> I'm sorry I haven't documented…

My suggestion: You can create a project here in the repository. There you can add notes for all your ideas and also issue reports and pull requests.

> …@driehle doing such great PRs…

Right! 👍 ❤️

Reply:
The reason I am pushing this currently is that some libraries have already migrated to laminas-code v4 and their updates are blocked because of laminas-form with v2 still sticking to laminas-code v3. :angel: